cmd/recovery-tool/volmgr/sfdisk.go

```diff
@@ -0,0 +1,260 @@
+// information as a laid out volume.
+func positionedVolumeFromDump(dump *SFDiskDeviceDump) (*gadget.LaidOutVolume, error) {
+	ptable := dump.PartitionTable
+
+	if ptable.Unit != "sectors" {
+		return nil, fmt.Errorf("cannot position partitions: unknown unit %q", ptable.Unit)
+	}
+
+	structure := make([]gadget.VolumeStructure, len(ptable.Partitions))
+	ps := make([]gadget.LaidOutStructure, len(ptable.Partitions))
+
+	for i, p := range ptable.Partitions {
+		info, err := filesystemInfo(p.Node)
+		if err != nil {
+			return nil, fmt.Errorf("cannot obtain filesystem information: %v", err)
+		}
+		if len(info.BlockDevices) < 1 {
+			continue
+		}
+		bd := info.BlockDevices[0]
+
+		structure[i] = gadget.VolumeStructure{
+			Name:       p.Name,
+			Size:       gadget.Size(p.Size) * sectorSize,
+			Label:      bd.Label,
+			Type:       p.Type,
+			Filesystem: bd.FSType,
+		}
+		ps[i] = gadget.LaidOutStructure{
+			VolumeStructure: &structure[i],
+			StartOffset:     gadget.Size(p.Start) * sectorSize,
+			Index:           i + 1,
+		}
+	}
+
+	pv := &gadget.LaidOutVolume{
+		Volume: &gadget.Volume{
+			ID:        ptable.ID,
+			Structure: structure,
+		},
+		Size:             gadget.Size(ptable.LastLBA),
+		SectorSize:       sectorSize,
+		LaidOutStructure: ps,
+	}
+
+	return pv, nil
+}
+
+func deviceName(name string, index int) string {
+	if len(name) > 0 {
+		last := name[len(name)-1]
+		if last >= '0' && last <= '9' {
+			return fmt.Sprintf("%sp%d", name, index)
+		}
+	}
+	return fmt.Sprintf("%s%d", name, index)
+}
+
+// buildPartitionList builds a list of partitions based on the current device contents
+// and gadget structure list, in sfdisk dump format. Return a partitioning description
+// suitable for sfdisk input and a role to device node map.
+func buildPartitionList(ptable *SFDiskPartitionTable, pv *gadget.LaidOutVolume, usedPartitions []bool) (*bytes.Buffer, map[string]string) {
+	buf := &bytes.Buffer{}
+	deviceMap := map[string]string{}
+
+	// Write partition data in sfdisk dump format
+	fmt.Fprintf(buf, "label: %s\nlabel-id: %s\ndevice: %s\nunit: %s\nfirst-lba: %d\nlast-lba: %d\n\n",
+		ptable.Label, ptable.ID, ptable.Device, ptable.Unit, ptable.FirstLBA, ptable.LastLBA)
+
+	for _, p := range ptable.Partitions {
+		fmt.Fprintf(buf, "%s : start=%12d, size=%12d, type=%s, uuid=%s", p.Node, p.Start,
+			p.Size, p.Type, p.UUID)
+		if p.Name != "" {
+			fmt.Fprintf(buf, ", name=%q", p.Name)
+		}
+		fmt.Fprintf(buf, "\n")
+	}
+
+	// Add missing partitions
+	for i, p := range pv.LaidOutStructure {
+		s := p.VolumeStructure
+		// Skip partitions that are already in the volume
+		if usedPartitions[i] {
+			continue
+		}
+		// Skip MBR structure
+		if s.Type == "mbr" {
+			continue
+		}
+		// Can we use the index here? Get the largest existing partition number and
+		// build from there could be safer if the disk partitions are not consecutive
+		// (can this actually happen in our images?)
+		node := deviceName(ptable.Device, p.Index)
```

Comment:
Took the simple approach but unsure if our image creation tools guarantee this index-to-partition-number parity.